engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/validation/StraightThroughProcessingLoopValidator.java

```diff
@@ -0,0 +1,222 @@
+      potentialLoop.addLast(element);
+      Either<List<ExecutableFlowNode>, ?> isPartOfLoop = Either.right(null);
+      for (final ExecutableSequenceFlow outgoing : element.getOutgoing()) {
+        isPartOfLoop = checkForStraightThroughProcessingLoop(potentialLoop, outgoing.getTarget());
+        if (isPartOfLoop.isLeft()) {
+          break;
+        }
+      }
+
+      if (isPartOfLoop.isRight()) {
+        potentialLoop.remove(element);
+      }
+      return isPartOfLoop;
+
+    } else {
+      // No loops have been found. We can return an Either.Right to indicate success.
+      return Either.right(null);
+    }
+  }
+
+  /**
+   * This method will decide if we are in a loop or not. It does so by checking if the potentialLoop
+   * list contains the element we are currently checking. It also has an extra check to make sure we
+   * don't detect loops between just gateways. As of the time of writing it is unclear whether we
+   * should reject these or not.
+   *
+   * @param potentialLoop the list of elements in a potential loop
+   * @param element the element we are currently checking
+   * @return a boolean indicating if we are in a loop or not
+   */
+  private static boolean foundLoop(
+      final LinkedList<ExecutableFlowNode> potentialLoop, final ExecutableFlowNode element) {
+    final var isLoopOfOnlyWhitelistedElements =
+        potentialLoop.stream()
+            .map(AbstractFlowElement::getElementType)
+            .allMatch(WHITELISTED_ELEMENT_TYPES::contains);
+    return potentialLoop.contains(element) && !isLoopOfOnlyWhitelistedElements;
+  }
+
+  /**
+   * Create a descriptive failure message which will be part of the rejection message. It mentions
+   * which resource and process contain the issue. It also lists the chain of element ids that are
+   * looping.
+   *
+   * @param resource the resource file that we've validated for loops
+   * @param executableProcess the process that we've validated for loops
+   * @param loopingElements a list of looping elements. The first and last element of this list will
+   *     be the same.
+   * @return a descriptive failure message
+   */
+  private static String createFailureMessage(
+      final DeploymentResource resource,
+      final ExecutableProcess executableProcess,
+      final List<ExecutableFlowNode> loopingElements) {
+    final List<String> loopingElementIds =
+        loopingElements.stream()
+            .map(AbstractFlowElement::getId)
+            .map(BufferUtil::bufferAsString)
+            .toList();
+
+    final StringWriter writer = new StringWriter();
+    writer.write(
+        String.format(
+            "`%s`: - Process: %s",
+            resource.getResourceName(), BufferUtil.bufferAsString(executableProcess.getId())));
+    writer.write("\n");
+    writer.write(
+        String.format(
+            "    - ERROR: Processes are not allowed to contain a straight-through processing loop: %s",
```

Comment:
Was thinking whether we want to add that at least an element that is a wait state needs to be included?
```suggestion
            "    - ERROR: Processes are not allowed to contain a straight-through processing loop without a single implemented task or intermediate event: %s",
```

Reply:
I don't like specifying undefined and manual tasks here to be honest. One reason for this is that the loop could also contain gateways, e.g.
<img width="620" alt="image" src="https://github.com/camunda/zeebe/assets/5787702/64409fd2-a99e-4921-9b25-735e18b23c5c">

This might be confusing for users. Another reason is that we'd have to change this message if/when we decide to add other elements to this validation.